File: app/src/main/java/com/v2/desafionubank/ui/BaseMain.java
package com.v2.desafionubank.ui;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.RelativeLayout;

import com.v2.desafionubank.ChargebackApplication;
import com.v2.desafionubank.R;
import com.v2.desafionubank.di.component.ActivityComponent;
import com.v2.desafionubank.di.component.DaggerActivityComponent;
import com.v2.desafionubank.di.module.ActivityModule;
import com.v2.desafionubank.model.ProcessRequest;
import com.v2.desafionubank.model.ResponseError;
import com.v2.desafionubank.ui.activity.ErrorActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import rx.functions.Action1;


/**
 * Created by CaioSChristino on 05/02/18.
 */

public abstract class BaseMain extends AppCompatActivity {
    protected final String ERROR_MESSAGE = "error_message";
    @Inject
    NavigationManager mNavigationManager;

    @BindView(R.id.progress)
    @Nullable
    RelativeLayout mProgressBar;

    ActivityComponent mActivityComponent;

    public abstract @LayoutRes
    int layoutResID();

    public abstract void initViews();

    public abstract NavigationManager.BackFragmentNavigationListener setBackNavigationListener();

    private ActivityComponent getActivityComponent() {
        if (mActivityComponent == null) {
            mActivityComponent = DaggerActivityComponent.builder()
                    .activityModule(new ActivityModule(this))
                    .applicationComponent(ChargebackApplication.getApp().getApplicationComponent())
                    .build();
        }
        return mActivityComponent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(layoutResID());
        ButterKnife.bind(this);
        getActivityComponent().inject(this);
        mNavigationManager.init(getSupportFragmentManager());

        ((ChargebackApplication) getApplication())
                .getBus()
                .toObservable()
                .subscribe((Action1<? super Object>) object -> {
                    if (object instanceof ResponseError) {
                        ResponseError error = (ResponseError) object;
                        Intent intent = new Intent(this, ErrorActivity.class);
                        intent.putExtra(ERROR_MESSAGE, error.getError());
                        startActivity(intent);
                    } else if (object instanceof ProcessRequest) {
                        ProcessRequest process = (ProcessRequest) object;
                        if (mProgressBar != null) {
                            mProgressBar.setVisibility(process.isShowProcess() ? View.VISIBLE : View.GONE);
                        }
                    }
                });
        initViews();
    }

    public void openFragment(BaseFragment fragment) {
        mNavigationManager.pushFragment(fragment);
    }

    @Override
    public void onBackPressed() {
        mNavigationManager.popFragment(setBackNavigationListener());
    }
}

File: app/src/main/java/com/v2/desafionubank/ui/BaseMain.kt
package com.v2.desafionubank.ui

import android.content.Intent
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v7.app.AppCompatActivity
import android.view.View
import com.v2.desafionubank.ChargebackApplication
import com.v2.desafionubank.di.component.ActivityComponent
import com.v2.desafionubank.di.component.DaggerActivityComponent
import com.v2.desafionubank.di.module.ActivityModule
import com.v2.desafionubank.model.ProcessRequest
import com.v2.desafionubank.model.ResponseError
import com.v2.desafionubank.ui.activity.ErrorActivity
import kotlinx.android.synthetic.main.activity_main.*
import rx.functions.Action1
import javax.inject.Inject


/**
 * Created by csanchez on 19/04/2018.
 */

abstract class BaseMain : AppCompatActivity() {
    protected val ERROR_MESSAGE = "error_message"
    @Inject
    lateinit var mNavigationManager: NavigationManager

    internal var mActivityComponent: ActivityComponent? = null
    private val activityComponent: ActivityComponent
        get() {
            if (mActivityComponent == null) {
                mActivityComponent = DaggerActivityComponent.builder()
                        .activityModule(ActivityModule(this))
                        .applicationComponent(ChargebackApplication.app!!.applicationComponent)
                        .build()
            }
            return mActivityComponent as ActivityComponent
        }

    @LayoutRes
    abstract fun layoutResID(): Int

    abstract fun initViews()

    abstract fun setBackNavigationListener(): NavigationManager.BackFragmentNavigationListener

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layoutResID())

        activityComponent.inject(this)
        mNavigationManager.init(supportFragmentManager)
        (application as ChargebackApplication)
                .bus!!
                .toObservable()
                .subscribe(Action1 { item ->
                    if (item is ResponseError) {
                        val error = item as ResponseError
                        val intent = Intent(this, ErrorActivity::class.java)
                        intent.putExtra(ERROR_MESSAGE, error.error)
                        startActivity(intent)
                    } else if (item is ProcessRequest) {
                        val request = item as ProcessRequest
                        progress?.setVisibility(if (request.isShowProcess) View.VISIBLE else View.GONE)
                    }
                })
        initViews()
    }

    fun openFragment(fragment: BaseFragment) {
        mNavigationManager!!.pushFragment(fragment)
    }

    override fun onBackPressed() {
        mNavigationManager!!.popFragment(setBackNavigationListener())
    }
}